nixos/modules/services/networking/netbird.nix

```diff
@@ -1,171 +1,513 @@
+            for binary in netbird ${optionalString cfg.ui.enable "netbird-ui"} ; do
+              ln -s "${client.wrapper}/bin/$binary-${name}" "$out/bin/$binary"
+            done
+          ''))
+        ];
+      }
+    ))
+    {
+      boot.extraModulePackages = optional
+        (cfg.clients != { } && (versionOlder kernel.version "5.6"))
+        kernelPackages.wireguard;
 
-      environment.systemPackages = [ cfg.package ];
+      environment.systemPackages =
+        toClientList (client: client.wrapper)
+        # omitted due to https://github.com/netbirdio/netbird/issues/1562
+        #++ optional (cfg.clients != { }) cfg.package
+        # omitted due to https://github.com/netbirdio/netbird/issues/1581
+        #++ optional (cfg.clients != { } && cfg.ui.enable) cfg.ui.package
+      ;
 
-      networking.dhcpcd.denyInterfaces = attrNames cfg.tunnels;
+      networking.dhcpcd.denyInterfaces = toClientList (client: client.interface);
+      networking.networkmanager.unmanaged = toClientList (client: "interface-name:${client.interface}");
 
-      systemd.network.networks = mkIf config.networking.useNetworkd (
-        mapAttrs'
-          (
-            name: _:
-            nameValuePair "50-netbird-${name}" {
-              matchConfig = {
-                Name = name;
-              };
-              linkConfig = {
-                Unmanaged = true;
-                ActivationPolicy = "manual";
-              };
-            }
-          )
-          cfg.tunnels
-      );
+      networking.firewall.allowedUDPPorts = concatLists (toClientList (client: optional client.openFirewall client.port));
 
-      systemd.services =
-        mapAttrs'
-          (
-            name:
-            { environment, stateDir, ... }:
-            nameValuePair "netbird-${name}" {
-              description = "A WireGuard-based mesh network that connects your devices into a single private network";
+      systemd.network.networks = mkIf config.networking.useNetworkd (toClientAttrs (client:
+        nameValuePair "50-netbird-${client.interface}" {
+          matchConfig = {
+            Name = client.interface;
+          };
+          linkConfig = {
+            Unmanaged = true;
+            ActivationPolicy = "manual";
+          };
+        }
+      ));
 
-              documentation = [ "https://netbird.io/docs/" ];
+      environment.etc = toClientAttrs (client: nameValuePair "${client.name}/config.d/50-nixos.json" {
+        text = builtins.toJSON client.config;
+        mode = "0444";
+      });
 
-              after = [ "network.target" ];
-              wantedBy = [ "multi-user.target" ];
+      systemd.services = toClientAttrs (client: nameValuePair client.name {
+        description = "A WireGuard-based mesh network that connects your devices into a single private network";
 
-              path = with pkgs; [ openresolv ];
+        documentation = [ "https://netbird.io/docs/" ];
 
-              inherit environment;
```

Comment:
This should be added again, currently the environment is not passed to the service at all. So most of the default settings will have no effect. (the exports from the preStart scripts don't automatically pass along to the main service)

Reply:
Never mind, ignore this. I overlooked the make-wrapper set-default that you used.